### DISCUSSION TITLE
output samples without sample rate?

### DISCUSSION THREAD

**Discussion by reppr** [pschatzmann/arduino-audio-tools#445] [Q&A] [answered]
I am new to this wonderful world of the arduino-audio-tools, a bit overwhelmed and an absolute beginner regarding i2s. So my very first questions will probably be quite stupid, please be patient witch me ;)

I am working on a program that generates audio (and light) patterns on ESP32 boards. Currently i use the
ESP DACs for testing but would like to change to external i2s DAC hardware now. My software does not work in a constant sample rate, but delivers samples, "when they are ready". So for a very much simplified example let's look at a square wave, that would just give a high sample value when it goes high and then a low value when it goes low, nothing in between.

My question is, if i2s (or the audio tools) can be used like this, or if I have to generate 'normal' sample data in a constant sample rate first?

Thank you for this wonderful software :)

Robert

**Comment by pschatzmann** [comment, marked as answer]
Yes, all audio output is based on a defined sample rate. However you do not need to take care of this for your case. 

I suggest that you just implement a subclass of SoundGenerator that provides the actual value that you want to output.
Have a look at how e.g. the [NoiseGenerator](https://github.com/pschatzmann/arduino-audio-tools/blob/main/src/AudioEffects/SoundGenerator.h) is implemented.

Here is an [example sketch](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-generator-i2s/streams-generator-i2s.ino) that uses an SoundGenerator
